pkg/generator/password/password.go

```diff
@@ -0,0 +1,129 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package password
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/sethvargo/go-password/password"
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	genv1alpha1 "github.com/external-secrets/external-secrets/apis/generators/v1alpha1"
+)
+
+type Generator struct{}
+
+const (
+	defaultLength      = 24
+	defaultSymbolChars = "~!@#$%^&*()_+`-={}|[]\\:\"<>?,./"
+	digitFactor        = 0.25
+	symbolFactor       = 0.25
+
+	errNoSpec    = "no config spec provided"
+	errParseSpec = "unable to parse spec: %w"
+	errGetToken  = "unable to get authorization token: %w"
+)
+
+type generateFunc func(
+	len int,
+	symbols int,
+	symbolCharacters string,
+	digits int,
+	noUpper bool,
+	allowRepeat bool,
+) (string, error)
+
+func (g *Generator) Generate(ctx context.Context, jsonSpec *apiextensions.JSON, kube client.Client, namespace string) (map[string][]byte, error) {
+	return g.generate(
+		jsonSpec,
+		generateSafePassword,
+	)
+}
+
+func (g *Generator) generate(jsonSpec *apiextensions.JSON, passGen generateFunc) (map[string][]byte, error) {
+	if jsonSpec == nil {
+		return nil, fmt.Errorf(errNoSpec)
+	}
+	res, err := parseSpec(jsonSpec.Raw)
+	if err != nil {
+		return nil, fmt.Errorf(errParseSpec, err)
+	}
+	symbolCharacters := defaultSymbolChars
+	if res.Spec.SymbolCharacters != nil {
+		symbolCharacters = *res.Spec.SymbolCharacters
+	}
+	passLen := defaultLength
+	if res.Spec.Length > 0 {
+		passLen = res.Spec.Length
+	}
+	digits := int(float32(passLen) * digitFactor)
+	if res.Spec.Digits != nil {
+		digits = *res.Spec.Digits
+	}
+	symbols := int(float32(passLen) * symbolFactor)
+	if res.Spec.Symbols != nil {
+		symbols = *res.Spec.Symbols
+	}
+	pass, err := passGen(
+		passLen,
+		symbols,
+		symbolCharacters,
+		digits,
+		res.Spec.NoUpper,
+		res.Spec.AllowRepeat,
+	)
+	if err != nil {
+		return nil, err
+	}
+	return map[string][]byte{
+		"password": []byte(pass),
+	}, nil
+}
+
+func generateSafePassword(
+	passLen int,
+	symbols int,
+	symbolCharacters string,
+	digits int,
+	noUpper bool,
+	allowRepeat bool,
+) (string, error) {
+	gen, err := password.NewGenerator(&password.GeneratorInput{
+		Symbols: symbolCharacters,
+	})
+	if err != nil {
+		return "", err
+	}
+	return gen.Generate(
+		passLen,
+		digits,
+		symbols,
+		noUpper,
+		allowRepeat,
+	)
+}
+
+func parseSpec(data []byte) (*genv1alpha1.Password, error) {
+	var spec genv1alpha1.Password
+	err := yaml.Unmarshal(data, &spec)
+	return &spec, err
+}
```

Comment:
broken manifests like
```
dataFrom:
- generator: #mistake on generator versus generatorRef
     apiVersion: generators.external-secrets.io/v1alpha1
     kind: Password
     name: "foo" 
```
Would still lead to a successful parse with an empty defaulted spec. For the other generators this isn't a problem because it would cause other errors to happen along the road, but for passwords it has the very weird behavior to take the generators default- which is really hard to troubleshoot (as it's a successful operation).

Is there any way we could forcefully check the `spec` is set, even if empty? This would prevent a lot of errors that are unintentional, IMO.

Reply:
good point; I'll take a look at verifying the ExternalSecret inline generator resource.

Reply:
as discussed, we'll drop inline generators :heavy_check_mark: